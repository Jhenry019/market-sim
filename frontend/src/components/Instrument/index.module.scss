@import '../../scss/root';
@import '../../scss/mixins/instrument.scss';
@import '../../scss/mixins//pill.scss';


.instrument {
  @include instrument;

  cursor: pointer;
}

.selected { background-color: var(--black-core); }

.pill {
  @include pill;
}

/** 
* For styling an instrument's price changes.
*/
.flat { background-color: var(--gray-core); }
.up { background-color: var(--green-core); }
.down { background-color: var(--red-core); }