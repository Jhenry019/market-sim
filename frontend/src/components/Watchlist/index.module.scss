.watchlist {
  margin-top: 3rem;
  color: #fff;
  background: #1b1b1b;
  border-radius: 4px;
}