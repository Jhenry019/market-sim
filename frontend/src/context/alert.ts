import React from 'react';


export const AlertContext = React.createContext({
  Type: "success",
  message: "",
  sendMessage: (mssg: string, Type: "success" | "error") => {}
});