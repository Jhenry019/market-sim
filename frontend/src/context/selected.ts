import React from 'react';


export const SelectedContext = React.createContext({
  symbol: '',
  nShares: 0,
  setSelected: (_: any) => {}
});