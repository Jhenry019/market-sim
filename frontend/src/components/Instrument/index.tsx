import React, { useContext } from 'react';
import classNames from '../../utils/classNames';
import usePrice from '../../hooks/usePrice';
import direction from '../../utils/priceDirection';
import { SelectedContext } from '../../context/selected';
import styles from './index.module.scss';


export interface InstrumentProps {
  /** The ticker symbol identifying the instrument */
  symbol: string;
  /** Indicates whether the ticker is being previewed */
  isSelected: boolean;
  /** The selector used in testing environments */
  dataTestId?: string;
}

export default function Instrument({
  symbol,
  isSelected,
  dataTestId
}: InstrumentProps): JSX.Element 
{ 
  const {price} = usePrice(symbol); 
  const DIRECTION = direction(price.change);
  const selected = useContext(SelectedContext);

  const changeSelectedInstrument = () => {
    selected.setSelected({
      ...selected, 
      nShares: 0, 
      symbol
    });
  };

  const priceMarkup = price.current && (
    <span>${price.current}</span>
  );

  return (
    <li 
      className={classNames(
        styles.instrument,
        isSelected && styles.selected
      )} 
      onClick={changeSelectedInstrument}
      data-testid={dataTestId}
    >
      <span>{symbol}</span>
      <div>
        <span
          id={`${symbol}--price`} 
          className={classNames(
            styles.pill,
            DIRECTION.flat && styles.flat,
            DIRECTION.up && styles.up,
            DIRECTION.down && styles.down
          )}
        >
          {priceMarkup}
        </span>
      </div>
    </li>
  );
}