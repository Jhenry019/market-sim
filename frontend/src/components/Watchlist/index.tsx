import React, { useContext } from 'react';
import Instrument, { InstrumentProps } from '../Instrument';
import { SelectedContext } from '../../context/selected';
import styles from './index.module.scss';


interface WatchlistProps {
  /** A list of instruments */ 
  list: Array<InstrumentProps>
  /** A selector used for testing purposes */
  dataTestId?: string;
}

export default function Watchlist({
  list,
  dataTestId
}: WatchlistProps): JSX.Element 
{
  const selected = useContext(SelectedContext);
  const instruments = list.map(instrument => (
		<Instrument 
			key={instrument.symbol} 
			symbol={instrument.symbol} 
			isSelected={instrument.symbol === selected.symbol}
      dataTestId={instrument.dataTestId}
		/>
	));

  return (
    <div className="col-md-4 mx-auto">
      <div 
        className={styles.watchlist} 
        data-testid={dataTestId}
      >
        <h4 className="text-center pt-3 pb-5">
          Watchlist
        </h4>
        <ul className="list-group px-3 pb-3">
          {instruments}
        </ul>
      </div>
    </div>
  );
}