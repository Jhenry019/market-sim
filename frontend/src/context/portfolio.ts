import React from 'react';
import { PortfolioProps } from '../components/Portfolio';


export const PortfolioContext = React.createContext({
  cash: undefined,
  equity: undefined,
  positions: [],
  updatePortfolio: (_: PortfolioProps) => {}
});