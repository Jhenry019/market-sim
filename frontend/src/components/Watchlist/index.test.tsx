import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { InstrumentProps } from '../Instrument';
import { SelectedContext } from '../../context/selected';
import Watchlist from '.';


const instruments: Array<InstrumentProps> = [
  {symbol: 'ABC', isSelected: true, dataTestId: "abc"},
  {symbol: 'XYZ', isSelected: false, dataTestId: "xyz"}
];

const context = {
  symbol: 'ABC',
  nShares: 0,
  setSelected: jest.fn()
};

it("renders a watchlist containing two instruments", () => {
  render(
    <SelectedContext.Provider value={context}>
      <Watchlist 
        list={instruments} 
        dataTestId="watchlist"
      />
    </SelectedContext.Provider>
  );

  expect(screen.getByTestId('abc').classList)
    .toContain('selected');
    
  expect(screen.getByTestId('watchlist'))
    .toMatchSnapshot();
});

it("changes the selected instrument", () => {
  /**
   * When an instrument is clicked, it should
   * call the setSelected function (mocked)
   * which changes the selected instrument.
   */
  render(
    <SelectedContext.Provider value={context}>
      <Watchlist 
        list={instruments} 
        dataTestId="watchlist"
      />
    </SelectedContext.Provider>
  );

  fireEvent.click(screen.getByTestId("xyz"));

  expect(context.setSelected)
    .toBeCalledWith({...context, symbol: "XYZ"});
});