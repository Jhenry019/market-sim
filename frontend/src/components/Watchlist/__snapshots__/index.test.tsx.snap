// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a watchlist containing two instruments 1`] = `
<div
  class="watchlist"
  data-testid="watchlist"
>
  <h4
    class="text-center pt-3 pb-5"
  >
    Watchlist
  </h4>
  <ul
    class="list-group px-3 pb-3"
  >
    <li
      class="instrument selected"
      data-testid="abc"
    >
      <span>
        ABC
      </span>
      <div>
        <span
          class="pill"
          id="ABC--price"
        />
      </div>
    </li>
    <li
      class="instrument"
      data-testid="xyz"
    >
      <span>
        XYZ
      </span>
      <div>
        <span
          class="pill"
          id="XYZ--price"
        />
      </div>
    </li>
  </ul>
</div>
`;
